tests/parser.js

```diff
@@ -1,56 +1,157 @@
-
-  t.deepEqual([
-    { word: 'oh' },
-    SPACE,
-    { word: 'hi' },
-    SPACE,
-    { value: 'thing is quoted here' },
-    SPACE,
-    { hole: true },
-    SPACE,
-    { value: 1.0 },
-    SPACE,
-    { variable: 'var' }
-  ], parse('oh hi "thing is quoted here" _ 1.0 $var'));
+  t.deepEqual([ word('hi')], parse('hi'), 'simple word');
+  t.deepEqual([ variable('foo') ], parse('$foo'), 'simple var');
+
+  t.deepEqual(
+    ts()
+    .w_('oh')
+    .w_('hi')
+    .v_('thing is quoted here')
+    .h_()
+    .v_(1.0)
+    .variable('var').done(),
+    parse('oh hi "thing is quoted here" _ 1.0 $var')
+  );
+
+});
+
+test('whitespace', t => {
+  t.deepEqual(
+    ts().w_('hi').w('you').done(),
+    parse('hi    you'),
+    'white space gets collapsed'
+  );
+});
+
+test('words', t => {
+  // this exists to document *how* things are being parsed
+  // for documentation purposes
+  t.deepEqual(ts().w(',,,,').done(), parse(',,,,'));
+  t.deepEqual(ts().w('a,y').done(), parse('a,y'));
+  t.deepEqual(ts().w('a').h().done(), parse('a_'));
+  t.deepEqual(
+    ts().w('(').variable('a').w_(',').variable('b').w(')').done(),
+    parse('($a, $b)')
+  );
+  t.deepEqual(ts().w_('hi.').w('you').done(), parse('hi. you'));
+  t.deepEqual(ts().w('hi').v(0.1).done(), parse('hi.1'));
+  t.deepEqual(ts().w('w').v(1).done(), parse('w1'));
+  t.deepEqual(ts().w('w').v('').done(), parse('w""'));
+  t.deepEqual(ts().id('').done(), parse('#'));
+});
 
+test('FIXME quotes', t => {
+  // these are weird/ambiguous grammar rules that
+  // are currently being forgiven, but should error out
+  // because they dont make sense and make parsing the grammar
+  // more difficult in other languages
+  t.deepEqual(ts().w('w"').done(), parse('w"'));
+  t.deepEqual(ts().w('"').done(), parse('"'));
+  t.deepEqual(ts().w('w"a').done(), parse('w"a'));
 });
 
 test('issue#2', t => {
-  t.deepEqual([ { value: 0.5 } ], parse('0.5'));
-  t.deepEqual([ { value: 0.5 } ], parse('.5'));
-  t.deepEqual([ { value: 10 } ], parse('10'));
-  t.deepEqual([
-    { word: 'gorog' },
-    SPACE,
-    { word: 'is' },
-    SPACE,
-    { word: 'at' },
-    SPACE,
-    { value: 0.1 },
-    { word: ',' },
-    SPACE,
-    { value: 5 }
-  ], parse('gorog is at 0.1, 5'));
-  t.deepEqual([
-    { word: 'gorog' },
-    SPACE,
-    { word: 'is' },
-    SPACE,
-    { word: 'at' },
-    SPACE,
-    { value: 0.1 },
-    { word: ',' },
-    SPACE,
-    { value: 5 }
-  ], parse('gorog is at .1, 5'));
-})
+  const half = [ value(0.5) ];
+  t.deepEqual(half, parse('0.5'));
+  t.deepEqual(half, parse('.5'));
+  t.deepEqual([ value(10) ], parse('10'));
+
+  const gorog = ts()
+    .w_('gorog')
+    .w_('is')
+    .w_('at')
+    .v(0.1)
+    .w_(',')
+    .v(5)
+    .done();
+
+  t.deepEqual(gorog, parse('gorog is at 0.1, 5'));
+  t.deepEqual(gorog, parse('gorog is at .1, 5'));
+});
+
+///////////////////////////////////
+// ... Helpers defined below ... //
+///////////////////////////////////
```

Comment:
A little bit meta, but would it make sense to test the helpers?

```js
t.deepEqual([{ word: 'oh'}], ts().w('oh'))
t.deepEqual([{ word: 'oh'}, { word: ' '}], ts().w_('oh'))
```

Not sure if this adds noise or clarity.